File: go/Minimum_Cost_to_Hire_K_Workers.go
package main

import "sort"

func mincostToHireWorkers(q []int, w []int, k int) float64 {
	l := len(q)
	idx := make([]int, l)
	for i := 0; i < l; i++ {
		idx[i] = i
	}
	sort.Slice(idx, func(i, j int) bool {
		return q[idx[i]]*w[idx[j]] > q[idx[j]]*w[idx[i]]
	})
	wg := idx[k-1]
	ans := q[idx[k-1]] * w[wg]
	for i := 0; i < k-1; i++ {
		ans += q[idx[i]] * w[wg]
	}
	for i := k; i < l; i++ {
		nx := (ans/w[wg] - q[idx[i-k]] + q[idx[i]]) * w[idx[i]]
		if nx*q[wg] < ans*q[idx[i]] {
			ans, wg = nx, idx[i]
		}
	}
	return float64(ans) / float64(q[wg])
}

File: cpp/Minimum_Cost_to_Hire_K_Workers.cpp
#include <algorithm>
#include <array>
#include <queue>
#include <vector>
using namespace std;

class Solution {
public:
  double mincostToHireWorkers(vector<int> &q, vector<int> &w, int k) {
    const auto l = q.size();
    vector<array<double, 2>> v(l);
    for (auto i = 0; i < l; i++) {
      v[i] = {(double)w[i] / q[i], (double)q[i]};
    }
    sort(v.begin(), v.end());
    auto qlt = 0;
    priority_queue<int> pq;
    for (auto i = 0; i < k; i++) {
      qlt += v[i][1];
      pq.push(v[i][1]);
    }
    double ans = qlt * v[k - 1][0];
    for (auto i = k; i < l; i++) {
      pq.push(v[i][1]);
      qlt += v[i][1] - pq.top();
      pq.pop();
      ans = min(ans, qlt * v[i][0]);
    }
    return ans;
  }
};
